### PR TITLE
Backend challenge

### PR DESCRIPTION
**Pull request by rwillians**
**TL;DR:** see `backend/script.rb` file.

Ruby isn't my main programming language, so please let me know what you guys think and what can I improve.

Thanks!

---

#### Payment

- I've changed its constructor method signature so now it depends upon a `order` and a `payment_method` to be instantiated, which are the minimum required information;
- I've made `authorization_number`, `invoice` and `paid_at` writable. This change allowed me to move the `pay` method (renamed to `charge`) to a `PaymentService` in order to comply with single responsibility principle;

#### Invoice

- I'm using `fetch` to retrieve a value from `attributes` because I expect an error in case any of those information are missing;
- Created `InvoiceFactory`: `PaymentService` has the need to create a new instance of `Invoice`. IME, instantiating a new `Invoice` is something you'll need to change more often than you'd like, so I've choose depend upon a factory to do that job;

#### Order

- No more `overrides`, the order is instantiated with a `Customer` and it's `Address` (which will be used as shipping address and billing address);
- Also removed unused `payment` attribute. `Payment` already has a relation back to `Order` and whatever class that has access to `Order` doesn't necessarily should have access to `Payment` information;

#### OrderItem

- I've added `price` property. Although `Product` already has a price, we need this property here too, to represent the actual price to be paid (after we apply discounts);

#### Products

- Well, IMO `book`, `digital product`, `physical product` and `membership` are different levels of types of products. Using polymorphism I've created the following inheritance: 
    - We have `Product`, which is a normal, physical, product. Like a book;
    - We have a concept representation: `VirtualProduct`. It represents the concept of something you can buy but it's not physical;
        -  A `VirtualProduct` can be a `DigitalProduct`: which could be a e-book or a downloadable content, for example;
        - And a `VirtualProduct`could also be a `Membership` or any other type of service for that matter;
- About the book taxes: I'm used to handle this situation using the product's NCM to resolve it's proper taxes aliquot, but since it will not be covered here, I've added an `is_book` flag attribute to the product;
- Also add a `active` flag to `Membership` to indicate the status of the subscription;

#### Credit Card

- Moved `fetch_by_hashed` (and renamed to `find_by_hash`) to a `CreditRepository` in order to keep the single responsibility principle;

#### Customer

- Added `name` and `email`;

#### MembershipService

- Created a `MembershipService` to control activation and deactivation of `Membership`s;

#### Checkout

- A service responsible for receiving an `Order` and a `payment_method`, apply the sales rules against the `Order` instance and then return `Payment` information which will be necessary in order to charge this order;

#### EventDispatcher

- IME, a `StateMachine` would benefit better a more robust solution. But, since this is a simple solution, I've choose to use events to handle most flux; In order to do that, I've create this simple dispatcher;

#### Email

- Represents the email to be sent;

#### EmailFactory

- Abstracts the process of constructing an email. This layer could handle/delegate template rendering among other needs;

#### EmailProvider

- Represents the client of an email provider;

#### NotificationService

- Receives the other to notify a customer about some event (what, in this context, is called `notification_type`) and delegates takes to `email_factory` and `email_provider` to achieve that;

#### Sales::Service

- Receives an `Order` and a `payment_method`;
- Then, using `Checkout` it process the order and gets the `Payment` information instance;
- Then, using `PaymentService` it validates the `Payment` information trying to avoid frauds;
- And, finally dispatches an event `order_received`

#### Sales Rules concept

- This can be better explained seen the code:

```ruby
# We should centralize all sales rules the business requires into
# some kind of sales rules manager.
sales_rules = Sales::Rules::Manager.new do |manager|
  # Challenge's 4th requirement states we need do apply a discount of
  # $10 to digital products:
  discount = Sales::Discounts::FixedDiscount.new(10)

  # This discount should be applied only to digital products within
  # the order:
  action = Sales::Actions::ApplyDiscount.new(discount) do |items|
    items.select { |item| item.product.is_a? DigitalProduct }
  end

  # Sales rules are executed against an order, so we need to make
  # sure there are digital products within the order. We can do
  # this kind of validation using conditions, such as this one:
  conditions = [Sales::Conditions::HasDigitalProduct.new]

  # Now we put it all together and add it to the manager:
  manager << Sales::Rules::Rule.new(
      name: 'Apply discount for digital products',
      action: action,
      conditions: conditions
  )
end
```

#### Shipping

For some reason I quite don't understand, the `shipping label` should "know" if a product is free of taxes. (Perhaps this was a mistype and the intention was `Invoice` - or a fiscal coupon -  know a given item is free of taxes?)

Anyways:

- I've created a `Shipping::Label` class which will received the `Customer`'s name, `Address` and a list of the items to be shipped;
- Since each `Item` should know if itself is free of taxes, I've created an `Shipping::Item` class, which holds an `OrderItem` and the flag `taxes_free`;
- The `Shipping::Service` receives the call to `generate_label` and uses a `Taxes::Service` to check if a given item should be flagged as `taxes_free`; `generate_label` currently just returns the `Shipping::Label` instance, but it could use some other service in order to have it printed;

#### Taxes::Service

Any taxes related information required by the application should be handled by this service. This service should delegate calculations to more specific classes, for calculations such as:
- PIS
- COFINS
- ICMS / ICMS ST
- Extimated total taxes

### CODE REVIEW

**Comment by hlegius** [comment]
Obrigado pelo PR! ;)
